Initial provider implementation for nerdctl (+finch)

estesp opened the pull request:
Adds implementation for a provider based on nerdctl. Several todos in the code but the core functionality of creating/deleting clusters is working and a simple application deployed works properly.

Fixes: #2317 

I don't like that it's a bit messy to pass around `binaryName` to support finch alongside nerdctl. Clearly users could alias finch -> nerdctl and this code could be simplified but I would love to find a clean way to support any nerdctl wrapper/implementation if possible without requiring user action.

**(Updated the following on 14 Dec 2023)**
~~Some TODOs in the code around IPv6 support (coming in nerdctl 1.7.0 but isn't released yet, and would be good to support nerdctl < 1.7.0 as well) and restart policy (supported, but containers are exiting with 137 with `kind delete cluster` so the one-time restart is actually happening, causing a recreate of the same cluster to fail without user action.~~

This PR now relies on nerdctl 1.7.0 or above, and/or finch 1.0.1 or above. That solves the lack of IPv6 support, although I have not tested IPv6 yet. The delete cluster flow was updated to use stop/wait/rm and works properly now, and leaves the restart policy on create matching the Docker provider.

**Update on 7 Feb 2024**
Now added the initial CI testing matrix for `nerdctl` from #3408 ; carried that PR's commit and modified it with recent versions and not aliasing to `docker` and instead using the implementation from this PR with experimental provider set.

AkihiroSuda commented on pkg/cluster/internal/providers/nerdctl/provision.go:
```diff
@@ -0,0 +1,396 @@
+		// note: requires API v1.41+ from Dec 2020 in Docker 20.10.0
+		// this is the default with cgroups v2 but not with cgroups v1, unless
+		// overridden in the daemon --default-cgroupns-mode
+		// https://github.com/docker/cli/pull/3699#issuecomment-1191675788
```
Doesn't make a sense for Docker

estesp replied:
Removed in latest commit

AkihiroSuda commented on pkg/internal/runtime/runtime.go:
```diff
@@ -18,6 +18,8 @@ func GetDefault(logger log.Logger) cluster.ProviderOption {
 		logger.Warn("using docker due to KIND_EXPERIMENTAL_PROVIDER")
 		return cluster.ProviderWithDocker()
+	case "nerdctl", "finch":
+		return cluster.ProviderWithNerdctl(p)
```
Would it be possible to support `nerdctl.lima` too?

estesp replied:
Yes!

estesp replied:
Done

AkihiroSuda commented on pkg/internal/runtime/runtime.go:
```diff
@@ -18,6 +18,8 @@ func GetDefault(logger log.Logger) cluster.ProviderOption {
 	case "docker":
 		logger.Warn("using docker due to KIND_EXPERIMENTAL_PROVIDER")
 		return cluster.ProviderWithDocker()
+	case "nerdctl", "finch":
```
Could you add a CI?
ref:
- https://github.com/kubernetes-sigs/kind/pull/3408

estesp replied:
yeah, I saw that PR.. should definitely collaborate on getting this into CI. Wanted to make sure there weren't any glaring misses here before getting CI set up.

AkihiroSuda replied:
> what is finch?

Amazon's distribution of nerdctl wrapped in Lima
https://aws.amazon.com/jp/blogs/opensource/ready-for-flight-announcing-finch-1-0-ga/

BenTheElder replied:
Are there compatibility issues such that we need to expand the test matrix that far as both upstream nerdctl and the finch distro?

AkihiroSuda commented on pkg/cluster/internal/providers/nerdctl/provider.go:
```diff
@@ -0,0 +1,361 @@
+)
+
+// NewProvider returns a new provider based on executing `nerdctl ...`
+func NewProvider(logger log.Logger, binaryName string) providers.Provider {
```
Wondering if the `nerdctl` package can be merged into the `docker` package. and the `docker` package should have something like `NewProviderWithVariant()` that takes a custom Docker-like binary path

estesp replied:
They are close; I also see some of the changes I had to make to a few capabilities around `inspect` in this PR are fixed (at least one around `Names` JSON object in container inspect) in recent PRs to nerdctl. There are also inconsistencies in a few other Go template responses that I worked around with `index`; probably could be fixed in nerdctl as it is essentially Docker compatibility mismatches at the moment.

AkihiroSuda replied:
Those inconsistencies should have been fixed in nerdctl v1.7
- https://github.com/containerd/nerdctl/pull/2594
- https://github.com/containerd/nerdctl/pull/2619

estesp replied:
Very nice; so the options are 1) try and use the docker provider with "binary name customization" and require nerdctl 1.7.0 or greater, or 2) create this provider and assume that over time it could be deprecated when any/all differences are resolved. Interested to hear from the kind maintainers what they would prefer. I think both are potentially reasonable, but obviously there is code duplication in creating yet-another-provider.

BenTheElder replied:
While I appreciate that nerdctl considers incompatibilities to be bugs and that there should be minimal skew, eventually we're bound to run into another incompatibility when developing the docker provider and need to work around these differences until a nerdctl update is released.

It's not a big deal to request that incoming changes be ported between essentially two copies of the docker provider, and in this PR we already have an example of something that should only be tested for docker and not nerdctl, https://github.com/kubernetes-sigs/kind/pull/3429#discussion_r1406840993

We already need to ensure that any new features target podman as well, often the code is nearly the same but not quite.

estesp replied:
Agree; that makes sense to me.

AkihiroSuda commented:
> coming in nerdctl 1.7.0 but isn't released yet

released: https://github.com/containerd/nerdctl/releases/tag/v1.7.0

estesp commented:
> > coming in nerdctl 1.7.0 but isn't released yet
> 
> released: https://github.com/containerd/nerdctl/releases/tag/v1.7.0

Oops! Well, you can tell how well I'm paying attention! I'm testing using Finch 1.0 which hasn't updated to 1.7.0 yet.

yankay commented on pkg/cluster/internal/providers/nerdctl/provider.go:
```diff
@@ -0,0 +1,361 @@
+		// filter for nodes with the cluster label
+		"--filter", fmt.Sprintf("label=%s=%s", clusterLabelKey, cluster),
+		// format to include the cluster name
+		"--format", `{{range .Names}}{{println .}}{{end}}`,
```
HI @estesp 

It's awsome PR :-)

The `nerdctl ps -a --filter label=io.x-k8s.kind.cluster=kind --format '{{range .Names}}{{println .}}{{end}}'` can not work in nerdctl , but `nerdctl ps -a --filter label=io.x-k8s.kind.cluster=kind --format '{{.Names}}'` same with docker can work correctly in v1.7.0 :-)

estesp replied:
Fixed to match behavior from nerdctl 1.7.0 and above

yankay commented:
HI @kind-maintainers,  @estesp

would you please review the PR , and give some advice? 
Thank you very much ╰(*°▽°*)╯

aojea commented:
/ok-to-test

aojea commented:
naive question, how complex is to add one github action job to test this in CI and get some signal?

aojea commented on pkg/cluster/internal/providers/nerdctl/network_integration_test.go:
```diff
@@ -0,0 +1,83 @@
+	"sigs.k8s.io/kind/pkg/internal/integration"
+)
+
+func TestIntegrationEnsureNetworkConcurrent(t *testing.T) {
```
This test fails on the CI because the job pod does not contain the nerdctl binary

BenTheElder replied:
We probably don't need this test for nerdctl and eventually even docker won't need it as the docker bug/quirk we were working around was fixed recently.

BenTheElder replied:
https://github.com/moby/moby/issues/20648#issuecomment-1716146390 🎉

BenTheElder commented on pkg/cluster/internal/providers/nerdctl/provider.go:
```diff
@@ -0,0 +1,361 @@
+
+	// ensure the pre-requisite network exists
+	networkName := fixedNetworkName
+	if n := os.Getenv("KIND_EXPERIMENTAL_DOCKER_NETWORK"); n != "" {
```
I would prefer that we not support this in the nerdctl provider, I consider it to have been a mistake in the docker provider that has caused no end of headaches (e.g. users trying to run in host network (!))

estesp replied:
Sounds good; happy to remove this piece

estesp replied:
Done in latest commit

BenTheElder commented:
> Some TODOs in the code around IPv6 support (coming in nerdctl 1.7.0 but isn't released yet, and would be good to support nerdctl < 1.7.0 as well) and restart policy (supported, but containers are exiting with 137 with kind delete cluster so the one-time restart is actually happening, causing a recreate of the same cluster to fail without user action.

Is the user action to `rm` again?
If so, can we sleep-wait for the container to restart and then rm again within Delete, and later add a check to skip this if nerdctl is newer than some fixed release?

arivera-xealth commented:
+1 on this

estesp commented:
I'm thinking for simplicity to make nerdctl 1.7.0 the lowest supported version; given there has been no support until now, that seems reasonable to me, but curious if anyone has strong opinions. I also would like input on whether to simply allow the provider to fail if the user has an older version, or do version checks during provider initialization?

estesp commented:
> > Some TODOs in the code around IPv6 support (coming in nerdctl 1.7.0 but isn't released yet, and would be good to support nerdctl < 1.7.0 as well) and restart policy (supported, but containers are exiting with 137 with kind delete cluster so the one-time restart is actually happening, causing a recreate of the same cluster to fail without user action.
> 
> Is the user action to `rm` again? If so, can we sleep-wait for the container to restart and then rm again within Delete, and later add a check to skip this if nerdctl is newer than some fixed release?

I really don't understand why the container is exiting with error; although ending with `rm -f` goes straight to `SIGKILL` to the container. I've corrected this for now with a slightly slower but stable approach I assume would work elsewhere: `stop` + `wait` + `rm -f -v`; I also prepended that with updating the `restart-policy` to `no` because the container still seems to exit with error and that (at least for nerdctl) means a restart try, which I didn't dig into yet.

At least for now, this is mitigated in the latest pushed code in the PR and several create, delete, create, delete attempts properly kill and remove nodes and allow for clusters to be created and deleted properly.

estesp commented:
Fixed everything from prior comments and re-pushed. Tested with nerdctl 1.7.1 and finch 1.0.1.

Remaining item is to handle CI needed steps to properly test (maybe carry the existing PR?) nerdctl+kind. Any other comments welcome. At this point I did not try to add version checking to try and handle versions of nerdctl < 1.7.0; they will simply fail to work properly. I'm guessing this may be reasonable as there is no pre-existing use of nerdctl+KinD and making 1.7.0 and above a baseline is probably reasonable for developer tooling.

estesp commented:
@BenTheElder and other maintainers; I've now added the start of CI from the #3408 PR and modified it to use this new experimental provider and updated to the recent releases of the dependencies (nerdctl, containerd, etc.). Would love to see if we can get testing validated and move forward. Of course if there are other issues or problems that need resolved, hopefully I can be responsive and get those resolved ASAP.

estesp commented:
Similar to Podman CI, I had to add `continue-on-error` to the load image step because the code in KinD is currently hardcoded to run a docker client command to get the image ID. Added a TODO in the github action yaml to match the podman one, but need to solve this in the code to respect KIND_EXPERIMENTAL_PROVIDER

estesp commented:
Can anyone approve the workflows for this to see if the added nerdctl workflow tests pass in GH actions?

estesp commented:
/retest

BenTheElder commented:
> Can anyone approve the workflows for this to see if the added nerdctl workflow tests pass in GH actions?

They're already approved?

Sorry, last week was enhancements freeze so pretty much all my time went to PRR reviews https://www.kubernetes.dev/resources/release/#timeline

aojea is OOO currently.

I'm going to be working on a bug fix release related to #3510 and some other work but aim to come back to this.

> I'm thinking for simplicity to make nerdctl 1.7.0 the lowest supported version; [...]

SGTM

> [...] given there has been no support until now, that seems reasonable to me, but curious if anyone has strong opinions. I also would like input on whether to simply allow the provider to fail if the user has an older version, or do version checks during provider initialization?



The current behavior is to fallback to docker if we don't detect another working provider unless the user explicitly sets `KIND_EXPERIMENTAL_PROVIDER`, if another provider is considered stable (podman is close-ish) then we might change that ...

Detecting providers is supposed to be fast and cheap because it happens on every invocation unless explicitly set, we're just looking for binaries in path.

In the podman provider we version check during `create cluster` implementation, for a few reasons:
- we only do the expensive calls during the once-per-cluster invocation
- this command inherently happens before any of the others
- other commands like `kind delete` will work fine with older versions anyhow

I think we should keep that pattern for now.

> At least for now, this is mitigated in the latest pushed code in the PR and several create, delete, create, delete attempts properly kill and remove nodes and allow for clusters to be created and deleted properly.

sounds OK for now, long term would love to get reboot working on all the providers but it's been thorny for podman as well and not perfect under docker either.

---

re: image-loading, we've had an open issue to outline a better command that is multi-provider, would love some more input there. the intention is to leave `kind load docker-image` and `kind load image-archive` unchanged (confusing compat at this point given `kind load docker-image` will actually do docker => podman if you have both on a system currently 🤯)

estesp commented:
>> Can anyone approve the workflows for this to see if the added nerdctl workflow tests pass in GH actions?
>
> They're already approved?
>

Oh cool! when I first commented I had made another small fix and re-pushed and they were stuck waiting on approval--looks like someone did that and they ran successfully, so yay!

> Sorry, last week was enhancements freeze so pretty much all my time went to PRR reviews > https://www.kubernetes.dev/resources/release/#timeline
>
>aojea is OOO currently.

No worries at all; no rush on this. Hopefully it is in good enough shape for review and has basic CI now for nerdctl.

aojea commented on pkg/internal/runtime/runtime.go:
```diff
@@ -18,6 +18,8 @@ func GetDefault(logger log.Logger) cluster.ProviderOption {
 	case "docker":
 		logger.Warn("using docker due to KIND_EXPERIMENTAL_PROVIDER")
 		return cluster.ProviderWithDocker()
+	case "nerdctl", "finch", "nerdctl.lima":
```
`logger.Warn("using nerdctl due to KIND_EXPERIMENTAL_PROVIDER")`

estesp replied:
Good catch; repushed. Thanks!

aojea commented:
one nit, missing a log on the provider detection, the rest looks ok, the code is isolated and is almost the same with the `binaryName` parameter.

estesp commented:
@aojea fixed the missing logger; should be ready to go then!

aojea commented:
/lgtm
/approve

/hold

just if @BenTheElder wants to check something, we can unhold tomorrow if no answer

estesp commented:
@BenTheElder any thoughts/concerns?

aojea commented:
/hold cancel

lazy consensus :)

nimakaviani commented:
great to see this @estesp!

BenTheElder commented:
Thank you all!

yankay commented:
Thanks @estesp 🎉🎉🎉
